src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,404 @@
+            State.COMPLETED,
+            ActivityStateMachine::notifyCompleted)
+        .add(
+            State.STARTED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCELED,
+            State.CANCELED,
+            ActivityStateMachine::notifyCancellationFromEvent);
+  }
+
+  private final ExecuteActivityParameters parameters;
+
+  private final Functions.Proc2<Optional<Payloads>, Failure> completionCallback;
+
+  private long startedCommandEventId;
+
+  /**
+   * @param parameters attributes used to schedule an activity
+   * @param completionCallback one of ActivityTaskCompletedEvent, ActivityTaskFailedEvent,
+   *     ActivityTaskTimedOutEvent, ActivityTaskCanceledEvents
+   * @param commandSink sink to send commands
+   * @return an instance of ActivityCommands
+   */
+  public static ActivityStateMachine newInstance(
+      ExecuteActivityParameters parameters,
+      Functions.Proc2<Optional<Payloads>, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink) {
+    return new ActivityStateMachine(parameters, completionCallback, commandSink);
+  }
+
+  private ActivityStateMachine(
+      ExecuteActivityParameters parameters,
+      Functions.Proc2<Optional<Payloads>, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink) {
+    super(newStateMachine(), commandSink);
+    this.parameters = parameters;
+    this.completionCallback = completionCallback;
+    explicitEvent(ExplicitEvent.SCHEDULE);
+  }
+
+  public void createScheduleActivityTaskCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK)
+            .setScheduleActivityTaskCommandAttributes(parameters.getAttributes())
+            .build());
+  }
+
+  public void cancel() {
+    if (parameters.getCancellationType() == ActivityCancellationType.ABANDON) {
+      notifyCanceled();
+    } else {
+      explicitEvent(ExplicitEvent.CANCEL);
+    }
+  }
+
+  private void setStartedCommandEventId() {
+    startedCommandEventId = currentEvent.getEventId();
+  }
+
+  private void cancelCommandNotifyCancelled() {
+    cancelCommand();
+    if (parameters.getCancellationType() != ActivityCancellationType.ABANDON) {
+      notifyCanceled();
+    }
+  }
+
+  private void notifyCanceledIfTryCancel() {
+    if (parameters.getCancellationType() == ActivityCancellationType.TRY_CANCEL) {
+      notifyCanceled();
+    }
+  }
+
+  private void notifyCanceled() {
+    ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+    Failure canceledFailure =
+        Failure.newBuilder()
+            .setSource(JAVA_SDK)
+            .setCanceledFailureInfo(CanceledFailureInfo.getDefaultInstance())
+            .build();
+    ActivityFailureInfo activityFailureInfo =
+        ActivityFailureInfo.newBuilder()
+            .setActivityId(scheduleAttr.getActivityId())
```

Comment:
[Nit] these fields seem more specific to ActivityInfo vs. ActivityFailureInfo

Reply:
I don't understand this comment.

Reply:
We have some ActivityInfo-ish fields in `ActivityFailureInfo`. They are needed to find activity in which failure was happened.